app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt

```diff
@@ -0,0 +1,247 @@
+                                        return@runOnUiThread
+                                    }
+                                }
+                                availableServicesAdapter.add(ServiceInfoWrapper(serviceInfo))
+                            }
+                        }
+                    }
+                    this@DiscoverActivity.nsdManager!!.resolveService(service, resolver)
+                } else {
+                    Log.d(TAG, "Unknown Service name: " + service.serviceName)
+                }
+            }
+
+            override fun onServiceLost(service: NsdServiceInfo) {
+                // When the network service is no longer available.
+                // Internal bookkeeping code goes here.
+                Log.e(TAG, "Service lost: $service")
+                multicastReleaser.run()
+            }
+
+            override fun onDiscoveryStopped(serviceType: String) {
+                Log.i(TAG, "Discovery stopped: $serviceType")
+                multicastReleaser.run()
+            }
+
+            override fun onStartDiscoveryFailed(serviceType: String, errorCode: Int) {
+                Log.e(TAG, "Discovery failed: Error code: $errorCode")
+                nsdManager.stopServiceDiscovery(this)
+                multicastReleaser.run()
+            }
+
+            override fun onStopDiscoveryFailed(serviceType: String, errorCode: Int) {
+                Log.e(TAG, "Discovery failed: Error code: $errorCode")
+                nsdManager.stopServiceDiscovery(this)
+                multicastReleaser.run()
+            }
+        }
+        nsdManager.discoverServices(
+                serviceType, NsdManager.PROTOCOL_DNS_SD, discoveryListener
+        )
+    }
+
+    private fun connectToChild(address: String, port: Int, name: String) {
+        val i = Intent(applicationContext, ListenActivity::class.java)
+        val b = Bundle()
+        b.putString("address", address)
+        b.putInt("port", port)
+        b.putString("name", name)
+        i.putExtras(b)
+        startActivity(i)
+    }
+
+    companion object {
+        private const val PREF_KEY_CHILD_DEVICE_ADDRESS = "childDeviceAddress"
+        private const val PREF_KEY_CHILD_DEVICE_PORT = "childDevicePort"
+    }
+}
+
+internal class ServiceInfoWrapper(private val info: NsdServiceInfo) {
+    fun matches(other: NsdServiceInfo): Boolean {
+        return info.host == other.host && info.port == other.port
+    }
+
+    val address: String
+        get() = info.host.hostAddress
+    val port: Int
+        get() = info.port
+    val name: String
+        get() {
+            // If there is more than one service with the same name on the network, it will
+            // have a number at the end, but will appear as the following:
+            //   "ChildMonitor\\032(number)
+            // or
+            //   "ChildMonitor\032(number)
+            // Replace \\032 and \032 with a " "
+            var serviceName = info.serviceName
```

Comment:
this could should probably never have gone into the getter but into the constructor instead

Reply:
Moved it to an `init()`